### PR TITLE
Store interface address and pass to announce callback

### PR DESCRIPTION
**Pull request by rolznz**
This allows the announce callback to create A/AAAA records with the address of the interface that received the request.

This means that we can correctly advertise the IP of our service on the local network.

### CODE REVIEW

File: examples/announce.c [hunk 1]
```diff
@@ -17,17 +17,23 @@
...
+bool stop(void*);
+void callback(void*, int, char*, const struct rr_entry*);
+
 void sighandler(int signum)
 {
         char s[] = "SIGINT received, exiting ...\n";
 
-        write(fileno(stdout), s, sizeof(s));
+        ssize_t result = write(fileno(stdout), s, sizeof(s));
+        (void)result;
```
Comment (tguillem) [inline examples/announce.c:32]: Fixing this warning should be done in a separate commit.

File: examples/announce.c [hunk 2]
```diff
@@ -17,17 +17,23 @@
...
+#include <string.h>
 
 #include "microdns.h"
 #include "compat.h"
 
 volatile sig_atomic_t sigflag = 0;
 
+void sighandler(int);
+bool stop(void*);
+void callback(void*, int, char*, const struct rr_entry*);
```
Comment (tguillem) [inline examples/announce.c]: Why are you added these forward declarations ?

File: src/microdns.h [hunk 3]
```diff
@@ -72,7 +72,8 @@ struct mdns_hdr {
         uint16_t num_add_rr;
 };
 
-typedef void (*mdns_callback)(void*, int, const struct rr_entry *);
+typedef void (*mdns_listen_callback)(void*, int, const struct rr_entry *);
+typedef void (*mdns_announce_callback)(void*, int, char *, const struct rr_entry *);
```
Comment (tguillem) [inline src/microdns.h]: Why doing it only for the announce part ? The listen part could also need it, no ?

File: src/mdns.c [hunk 4]
```diff
@@ -410,10 +510,10 @@ mdns_entries_send(const struct mdns_ctx *ctx, const struct mdns_hdr *hdr, const
...
 
         for (size_t i = 0; i < ctx->nb_conns; ++i) {
-            r = sendto(ctx->conns[i].sock, (const char *) buf, n, 0,
-                    (const struct sockaddr *) &ctx->addr, ss_len(&ctx->addr));
-            if (r < 0)
-                return (MDNS_NETERR);
+                r = sendto(ctx->conns[i].sock, (const char *) buf, n, 0,
+                        (const struct sockaddr *) &ctx->addr, ss_len(&ctx->addr));
+                if (r < 0)
+                        return (MDNS_NETERR);
```
Comment (tguillem) [inline src/mdns.c]: tabulation instead of spaces here.

File: src/microdns.h (hunk 3 above)
Comment (tguillem) [reply src/microdns.h]: Instead of passing a char *, you should maybe use a struct like the following:

```
struct ip_source {
    int af;
    union {
        struct ipv4 {
            char   addr_str[INET_ADDRSTRLEN];
            struct in_addr addr;
        }
        struct ipv6 {
            char   addr_str[INET6_ADDRSTRLEN];
            struct in6_addr addr;
        }
    }
}
```

File: src/mdns.c [hunk 5]
```diff
@@ -89,9 +97,10 @@ mdns_is_interface_valuable(struct ifaddrs* ifa)
 }
 
 static int
-mdns_list_interfaces(multicast_if** pp_intfs, size_t* p_nb_intf, int ai_family)
+mdns_list_interfaces(multicast_if** pp_intfs, char ***pp_ip_addresses, size_t* p_nb_intf, int ai_family)
 {
-        struct ifaddrs *ifs;
+        char **ip_addresses = *pp_ip_addresses = NULL;
```
Comment (tguillem) [inline src/mdns.c]: confusing naming, it should be
char ***ppp_ip_addresses;
char **pp_ip_addresses;

File: src/mdns.c [hunk 6]
```diff
@@ -112,31 +121,75 @@ mdns_list_interfaces(multicast_if** pp_intfs, size_t* p_nb_intf, int ai_family)
...
+        if (ip_addresses == NULL) 
+                goto error;
+
         for (c = ifs; c != NULL; c = c->ifa_next) {
                 if (c->ifa_addr == NULL ||
                     c->ifa_addr->sa_family != ai_family ||
                     !mdns_is_interface_valuable(c))
                         continue;
                 memcpy(intfs, c->ifa_addr, sizeof(*intfs));
+
```
Comment (tguillem) [inline src/mdns.c]: stray space

File: src/mdns.c [hunk 7]
```diff
@@ -112,31 +121,75 @@ mdns_list_interfaces(multicast_if** pp_intfs, size_t* p_nb_intf, int ai_family)
...
+                        struct sockaddr_in *addr_in = (struct sockaddr_in *)c->ifa_addr;
+                        ip_address = inet_ntop( ai_family, &addr_in->sin_addr, 
+                                                    address_buf, sizeof( address_buf ) );
+                }
+                else
+                {
+                        struct sockaddr_in6 *addr_in = (struct sockaddr_in6 *)c->ifa_addr;
+                        ip_address = inet_ntop( ai_family, &addr_in->sin6_addr, 
+                                                    address_buf, sizeof( address_buf ) );
+                }
```
Comment (tguillem) [inline src/mdns.c:141]: You should check inet_ntop doesn't return NULL

File: src/mdns.c [hunk 8]
```diff
@@ -112,31 +121,75 @@ mdns_list_interfaces(multicast_if** pp_intfs, size_t* p_nb_intf, int ai_family)
...
+
+                *ip_addresses = strdup(ip_address);
```
Comment (tguillem) [inline src/mdns.c]: cf. my remark about using a struct to propagate the source address.
It should make it simpler: one less alloc.

File: src/mdns.c [hunk 9]
```diff
@@ -193,17 +247,22 @@ mdns_list_interfaces(multicast_if** pp_intfs, size_t* p_nb_intf, int ai_family)
         }
         if (nb_intf == 0) {
                 *pp_intfs = malloc(sizeof(*intfs));
+                *pp_ip_addresses = malloc(sizeof(char*));
```
Comment (tguillem) [inline src/mdns.c]: missing alloc check

File: src/mdns.c [hunk 10]
```diff
@@ -212,9 +271,46 @@ mdns_list_interfaces(multicast_if** pp_intfs, size_t* p_nb_intf, int ai_family)
...
+                                struct sockaddr_in* si = (struct sockaddr_in*) p_unicast->Address.lpSockaddr;
+                                ip_address = inet_ntop( ai_family, &(si->sin_addr), 
+                                                                address_buf, sizeof( address_buf ) );
+                        }
+                        else
+                        {
+                                struct sockaddr_in6* si = (struct sockaddr_in6*) p_unicast->Address.lpSockaddr;
+                                ip_address = inet_ntop( ai_family, &(si->sin6_addr), 
+                                                                address_buf, sizeof( address_buf ) );
+                        }
```
Comment (tguillem) [inline src/mdns.c:268]: Same missing inet_ntop check

**Review by tguillem — CHANGES_REQUESTED** [review]

File: examples/announce.c [hunk 11]
```diff
@@ -36,7 +42,7 @@ bool stop(void *cbarg)
         return (sigflag ? true : false);
 }
 
-void callback(void *cbarg, int r, const struct rr_entry *entry)
+void callback(void *cbarg, int r, char *ip_address, const struct rr_entry *entry)
```
Comment (chouquette) [inline examples/announce.c]: This parameter should be const

File: src/mdns.c [hunk 12]
```diff
@@ -112,31 +121,75 @@ mdns_list_interfaces(multicast_if** pp_intfs, size_t* p_nb_intf, int ai_family)
                 return (MDNS_ERROR);
         }
         *pp_intfs = intfs = malloc(sizeof(*intfs) * nb_if);
-        if (intfs == NULL) {
-                freeifaddrs(ifs);
-                return (MDNS_ERROR);
-        }
```
Comment (chouquette) [inline src/mdns.c]: It might be easier to read if you split the error refactoring in a separate commit

File: src/mdns.c [hunk 13]
```diff
@@ -112,31 +121,75 @@ mdns_list_interfaces(multicast_if** pp_intfs, size_t* p_nb_intf, int ai_family)
...
+                if (*ip_addresses == NULL)
+                        goto error;
+                ip_addresses++;
                 intfs++;
         }
         freeifaddrs(ifs);
         *p_nb_intf = nb_if;
         return (0);
+error:
+        while (ip_addresses != *pp_ip_addresses)
```
Comment (chouquette) [inline src/mdns.c]: Wouldn't it be simpler to use `nb_if` for this loop?

File: src/mdns.c [hunk 14]
```diff
@@ -112,31 +121,75 @@ mdns_list_interfaces(multicast_if** pp_intfs, size_t* p_nb_intf, int ai_family)
...
+mdns_list_interfaces(multicast_if** pp_intfs, char ***pp_ip_addresses, size_t* p_nb_intf, int ai_family)
 {
         multicast_if *intfs;
         *pp_intfs = intfs = malloc(sizeof(*intfs));
         if (intfs == NULL)
                 return (MDNS_ERROR);
         memset(intfs, 0, sizeof(*intfs));
+
+        *pp_ip_addresses = malloc(sizeof(char*));
+        if (pp_ip_addresses == NULL)
```
Comment (chouquette) [inline src/mdns.c]: You're not checking the pointer that was allocated

File: src/mdns.c [hunk 15]
```diff
@@ -112,31 +121,75 @@ mdns_list_interfaces(multicast_if** pp_intfs, size_t* p_nb_intf, int ai_family)
...
+        {
+                free(*pp_intfs);
+                return (MDNS_ERROR);
+        }
+        memset(pp_ip_addresses, 0, sizeof(char*));
```
Comment (chouquette) [inline src/mdns.c]: I might be missing something, but couldn't you just store `NULL` in `*pp_ip_addresses`?

File: src/mdns.c [hunk 16]
```diff
@@ -177,9 +231,9 @@ mdns_list_interfaces(multicast_if** pp_intfs, size_t* p_nb_intf, int ai_family)
                 res = malloc( size );
                 if (res == NULL)
                         return (MDNS_ERROR);
-                hr = GetAdaptersAddresses(ai_family, GAA_FLAG_SKIP_UNICAST |
-                                                    GAA_FLAG_SKIP_ANYCAST | GAA_FLAG_SKIP_DNS_SERVER,
-                                                    NULL, res, &size);
```
Comment (chouquette) [inline src/mdns.c]: Whitespace change

File: src/microdns.h [hunk 17]
```diff
@@ -72,7 +72,8 @@ struct mdns_hdr {
         uint16_t num_add_rr;
 };
 
-typedef void (*mdns_callback)(void*, int, const struct rr_entry *);
```
Comment (chouquette) [inline src/microdns.h:75]: I think the renaming should be in a separate commit, and that probably means you need to bump some versions since this is a breaking API change

**Review by chouquette — CHANGES_REQUESTED** [review]

File: src/microdns.h (hunk 17 above)
Comment (chouquette) [reply src/microdns.h:75]: We might worry about bumping the version out of this PR though, let's do it when this is merged

File: examples/announce.c (hunk 2 above)
Comment (rolznz) [reply examples/announce.c]: I was getting warnings for those too. I can move them to a separate commit.

File: examples/announce.c (hunk 1 above)
Comment (rolznz) [reply examples/announce.c:32]: Ok, sounds good.

File: src/microdns.h (hunk 3 above)
Comment (rolznz) [reply src/microdns.h]: The listen callback doesn't need it because it can read the IP address from the A/AAAA record (rr_entry) in the response packet.

Why store both the address string and the in_addr?

File: src/mdns.c (hunk 5 above)
Comment (rolznz) [reply src/mdns.c]: Will do, thanks

File: src/mdns.c (hunk 4 above)
Comment (rolznz) [reply src/mdns.c]: There seems to be inconsistent tabs/spacing already. A lot is indented 8 spaces and some are 4. Can I create a commit to change all indentation to 4 spaces to match VLC?

File: src/mdns.c (hunk 7 above)
Comment (rolznz) [reply src/mdns.c:141]: Good idea, thanks

File: src/mdns.c (hunk 8 above)
Comment (rolznz) [reply src/mdns.c]: Sounds good to me.

File: src/mdns.c (hunk 9 above)
Comment (rolznz) [reply src/mdns.c]: Will add :-)

File: examples/announce.c (hunk 11 above)
Comment (rolznz) [reply examples/announce.c]: Good point.

File: src/mdns.c (hunk 12 above)
Comment (rolznz) [reply src/mdns.c]: Ok :-)

File: src/mdns.c (hunk 13 above)
Comment (rolznz) [reply src/mdns.c]: Yes. Although I will have to set all the IP addresses to NULL when ip_addresses is allocated

File: src/mdns.c (hunk 14 above)
Comment (rolznz) [reply src/mdns.c]: Oops - I'll fix this. Thanks.

File: src/mdns.c (hunk 15 above)
Comment (rolznz) [reply src/mdns.c]: Yes, I just followed what was consistent with the lines above (line 183)

File: src/microdns.h (hunk 17 above)
Comment (rolznz) [reply src/microdns.h:75]: Sounds good to me.

File: src/mdns.c (hunk 15 above)
Comment (rolznz) [reply src/mdns.c]: Actually, see See contrib\src\microdns\0001-build-fix-getifaddrs-detection.patch in VLC
@tguillem can you answer this?

File: examples/announce.c (hunk 2 above)
Comment (tguillem) [reply examples/announce.c]: The correct fix is to set these functions static

File: src/microdns.h (hunk 3 above)
Comment (tguillem) [reply src/microdns.h]: To do like rr_entry. I'm wondering what is best way to expose addresses, the string or the struct in_addr.

File: src/mdns.c (hunk 4 above)
Comment (tguillem) [reply src/mdns.c]: 8 spaces is better for this project, otherwise the diff will be huge.

File: src/mdns.c (hunk 15 above)
Comment (tguillem) [reply src/mdns.c]: Oh, VLC patches should be merged in this project, I'll do a PR.
But I don't see the link with Hugo's comment ?

Comment (tguillem) [reply src/mdns.c]: OK, these patches are already merged but we didn't do any new release since. That is why there are still in VLC.

File: examples/announce.c (hunk 2 above)
Comment (rolznz) [reply examples/announce.c]: Nice, I should know that :-)

File: src/microdns.h (hunk 3 above)
Comment (rolznz) [reply src/microdns.h]: I'm happy with either, I think one should be chosen (Seems wasteful to have both when you can calculate in_addr from the address string?)

File: src/mdns.c (hunk 15 above)
Comment (rolznz) [reply src/mdns.c]: OK cool, thanks for checking.

Comment (rolznz) [reply src/mdns.c]: @tguillem oops - I copied the wrong VLC patch - it should be contrib\src\microdns\0002-fix-dummy-mdns_list_interfaces.patch. 

Rather than doing memset(pp_ip_addresses, 0, sizeof(char*)); do *pp_ip_addresses = NULL; 

But a few lines above it uses memset so I just followed the same logic to be consistent.

In your patch you changed it to memset:
-        **pp_intfs = NULL;
+        memset(intfs, 0, sizeof(*intfs));

File: src/microdns.h (hunk 3 above)
Comment (tguillem) [reply src/microdns.h]: So I would prefer keeping the struct in*_addr. That is what avahi is doing for example: https://avahi.org/doxygen/html/struct_avahi_address.html

Comment (rolznz) [reply src/microdns.h]: Ok, sounds good to me. Thanks :-)

File: src/mdns.c (hunk 4 above)
Comment (rolznz) [reply src/mdns.c]: Ok, so these lines were previously indented with 4 spaces (inconsistent with the rest of the file). This change fixed those lines to be indented with 8 spaces, so should be correct now. I didn't actually change those lines apart from the whitespace so I can revert it or remove the change to a separate commit.

**Comment by rolznz** [comment]
@chouquette @tguillem I've checked in a new version. When you have time, please let me know what you think 😁

File: src/ip_source.h [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+    unsigned int family;
+    union {
+        struct {
+            struct in_addr addr;
+        } ipv4;
+        struct {
+            struct in6_addr addr;
+        } ipv6;
+    };
+};
```
Comment (tguillem) [inline src/ip_source.h]: I would have put this new struct in the microdns.h file.

Also, ip_source is a good name for a variable for your usecase, but it's not generic enough for the name of this struct (We may want to use this same struct as a destination for example).

I propose mdns_ip.

**Review by tguillem — CHANGES_REQUESTED** [review]

File: src/ip_source.h (hunk 18 above)
Comment (rolznz) [reply src/ip_source.h]: Sounds fair enough to me. Thanks.

Comment (rolznz) [reply src/ip_source.h]: @tguillem I've made the requested changes.

**Review by tguillem — APPROVED** [review]

**Review by chouquette — APPROVED** [review]

**Comment by tguillem** [comment]
Pushed into master. Thanks a lot for your work.

**Comment by rolznz** [comment]
@tguillem thanks - can you let me know when version 0.0.11 has been consumed by VLC?

**Comment by tguillem** [comment]
0.1.0 has just been released with your changes, and I proposed the contrib/configure update for VLC on the mailing list.

**Comment by rolznz** [comment]
Great :-) Thanks Thomas!
